### ISSUE TITLE
Add options for encoding nested form and query params a la clj-http

### ISSUE THREAD

**Issue by wevre** [http-kit/http-kit#561]
I’m using http-kit to make client calls to Stripe, but the default handling of nested form params produces something different than what Stripe expects. Here is what I’ve found:

- clj-http, whose API http-kit is modeled after, provides four options, based on the value of `:multi-param-style`, each of which encodes `{:a [1 2 3]}` as follows:
  - `nil` — “a=1&a=2&a=3"
  - `:array` — “a[]=1&a[]=2&a[]=3”
  - `:indexed` — “a[0]=1&a[1]=2&a[2]=3"
  - `:comma-separated` — “a=1,2,3”

- http-kit does not offer all those options. For nested maps, http-kit will produce a chain of bracketed keys (a bracketed key path), but it bottoms out at a vector and outputs the same as the `nil` flavor above. This means maps inside of vectors are in trouble.
- Stripe needs the output to match `:indexed` and to handle both nested maps and nested vectors (treating the index as the key).
- There are two pertinent closed issues:
  - [№ 97](https://github.com/http-kit/http-kit/issues/97) — closed for inactivity.
  - [№ 301](https://github.com/http-kit/http-kit/issues/301) — the person who opened the issue (also trying to communicate with Stripe) ended up creating their own small library to take over encoding form params and then pass the result in as `:body`.

- I don’t want to add another dependency, the ideal would be for http-kit to offer some options. For now I made my own copy of `org.httpkit.client/nested-param`, tweaked it slightly, and used it to pre-process my form params before handing them over to http-kit (not as `:body`, still under `:form-params`). Nested maps and vectors are flattened into string keys comprising the bracketed key path paired with the original value, and http-kit treats them like any other non-nested form param (albeit now with a string key instead of a keyword key).

- For the above to work, `nested-param` (or, rather, my tweaked copy of it) can’t assume that keys will always be keywords or symbols and call `name` on them; if it came from a vector, the key will be an integer. So all together my changes are:

```clojure
(defn- name* [v]
  (if (instance? clojure.lang.Named v)
    (name v)
    v))

(defn de-nest
  "{:a {:b 1 :c [1 2 3]}} => {\"a[b]\" 1, \"a[c][0]\" 1, \"a[c][1]\" 2, \"a[c][2]\" 3}, etc."
  [params]
  ;; Copied from httpkit, modified to allow for indexing of vectors.
  ;; https://github.com/http-kit/http-kit/blob/1421944d06d2b25f48cebb4d1e7f24ae7ca9b814/src/org/httpkit/client.clj#L40
  (walk/prewalk
   (fn [d]
     (if (and (vector? d) (or (map? (second d)) (vector? (second d))))
                                               ;^^^^^^^^^^^^^^^^^^^^   this is new
       (let [[fk m] d]
         (reduce-kv (fn [m sk v]
         ;^^^^^^^^^  reduce-kv instead of reduce
                   (assoc m (str (name* fk) \[ (name* sk) \]) v))
                                 ;^^^^^         ^^^^^   also new
                 {} m))
       d))
   params))
```

With a few additional tweaks, it would be possible to support all the styles originally supported by clj-http. I'm more than happy to create a PR.

**Comment by ptaoussanis** [comment]
@wevre Hi Mike,

Thanks for the clear info, that's very helpful! Would be happy to take a look at a PR to add options to match `clj-http`! 👍 🙏

**Comment by wevre** [comment]
Is there a recommended Java version to use when running the tests? I'm currently running with OpenJDK Temurin-21.0.1+12 but not all the tests are compiling/passing.

**Comment by ptaoussanis** [comment]
@wevre To clarify: the tests fail on your system even as of the current master, or only after your changes?

The tests aren't unusually sensitive to JVM version as far as I'm aware, and are passing with at least OpenJDK 21 and 17.

Could you please share the failure you're seeing?

**Comment by ptaoussanis** [comment]
BTW I've just updated the CI tests to [cover](https://github.com/http-kit/http-kit/blob/79395f0c1e016b6e6da41273be7c3e16a757d2be/.github/workflows/main-tests.yml#L8) Corretto OpenJDK 17, 19, and 21 - all of which seem okay on the current master 👍

**Comment by wevre** [comment]
Is it possible to use httpkit as a `:local/root` dependency? It must be, right? 

I wanted to test the fork directly in my project to do some Stripe calls, but when I start up my system/server (I'm using Integrant) I get an error that:

```
{:type java.io.FileNotFoundException
   :message "Could not locate httpkit/with_channel__init.class, httpkit/with_channel.clj or httpkit/with_channel.cljc on classpath. Please check that namespaces with dashes use underscores in the Clojure file name."
   :at [clojure.lang.RT load "RT.java" 462]}
```

The more I think about this, I think it is an issue with Integrant and attempting to reload changed namespaces. But why does it think that the macro `org.httpkit.server/with-channel` is a namespace?

UPDATE: I found a with-channel namespace, under 'resources/clj-kondo.exports'. I'm not familiar with exactly what that is, but I nuked it and am able to now work with my fork as a `:root/local` deps. I tested `:nested-param-style` `:indexed` and it worked.
